src/pkg/p2p/preheat/job.go

```diff
@@ -0,0 +1,225 @@
+type Job struct{}
+
+// MaxFails of preheat job. Don't need to retry.
+func (j *Job) MaxFails() uint {
+	return 1
+}
+
+// MaxCurrency indicates no limitation to the concurrency of preheat job.
+func (j *Job) MaxCurrency() uint {
+	return 0
+}
+
+// ShouldRetry indicates no need to retry preheat job as it's just for a cache purpose.
+func (j *Job) ShouldRetry() bool {
+	return false
+}
+
+// Validate the parameters of preheat job.
+func (j *Job) Validate(params job.Parameters) error {
+	_, err := parseParamProvider(params)
+	if err != nil {
+		return err
+	}
+
+	_, err = parseParamImage(params)
+
+	return err
+}
+
+// Run the preheat process.
+func (j *Job) Run(ctx job.Context, params job.Parameters) error {
+	// Get logger
+	myLogger := ctx.GetLogger()
+
+	// Parse parameters, ignore errors as they have been validated already
+	p, _ := parseParamProvider(params)
+	pi, _ := parseParamImage(params)
+
+	// Print related info to log first
+	myLogger.Infof(
+		"Preheating image '%s:%s' to the target preheat provider: %s %s:%s",
+		pi.ImageName,
+		pi.Tag,
+		p.Vendor,
+		p.Name,
+		p.Endpoint,
+	)
+
+	// Get driver factory for the given provider
+	fac, ok := pr.GetProvider(p.Vendor)
+	if !ok {
+		err := errors.Errorf("No driver registered for provider %s", p.Vendor)
+		myLogger.Error(err)
+		return preheatJobRunningError(err)
+	}
+
+	// Construct driver
+	d, err := fac(p)
+	if err != nil {
+		myLogger.Error(err)
+		return preheatJobRunningError(err)
+	}
+
+	myLogger.Infof("Get preheat provider driver: %s", p.Vendor)
+
+	// Start the preheat process
+	// First, check the health of the provider
+	h, err := d.GetHealth()
+	if err != nil {
+		return preheatJobRunningError(err)
+	}
+
+	if h.Status != pr.DriverStatusHealthy {
+		return preheatJobRunningError(errors.Errorf("unhealthy target preheat provider"))
+	}
+
+	myLogger.Infof("Check health of preheat provider instance: %s", pr.DriverStatusHealthy)
+
+	// Then send the preheat requests to the target provider.
+	st, err := d.Preheat(pi)
+	if err != nil {
+		return preheatJobRunningError(err)
+	}
+
+	myLogger.Info("Sending preheat request is successfully done")
+
+	// For some of the drivers, e.g: Kraken, the returned status of preheating request contains the
+	// final status info. No need to loop check the status.
+	if st.Status == provider.PreheatingStatusSuccess {
+		myLogger.Info("Preheating is completed")
+		return nil
+	}
+
+	myLogger.Info("Start to loop check the preheating status until it's ready or timeout(30m)")
+	// If process is not completed, loop check the status until it's ready.
+	tk := time.NewTicker(checkInterval)
+	defer tk.Stop()
+
+	tm := time.NewTimer(checkTimeout)
+	defer tm.Stop()
+
+	for {
+		select {
+		case <-tk.C:
+			s, err := d.CheckProgress(st.TaskID)
+			if err != nil {
+				myLogger.Error(err)
+				return preheatJobRunningError(err)
+			}
+
+			// Finished
+			if s.Status == provider.PreheatingStatusSuccess {
```

Comment:
Status will nerver be PreheatingStatusFail?

Reply:
You should check your Kraken driver logic. This is a specific logic for Kraken driver.

In the Kraken driver comments, if preheat is failed, an error will be returned. Otherwise, it will definitely return success.

Reply:
Anyway, enhance the logic for more flexible cases in the future

Reply:
it is not related to Kraken but the design. we have designed two final status PreheatingStatusSuccess and PreheatingStatusFail. so we should check these two status both here.

Reply:
I totally understood. So far, this specific logic actually is only for Kraken case as Dragonfly will never return other statuses(it will return `pending`). Kraken will return error or non-error and Success.

Anyway, enhance the logic to support the original design.

Reply:
Dragonfly: error or non-error & pending status

```go
bytes, err := client.GetHTTPClient(dd.instance.Insecure).Post(url, dd.getCred(), preheatingImage, nil)
	if err != nil {
		return nil, err
	}

	result := &dragonflyPreheatCreateResp{}
	if err := json.Unmarshal(bytes, result); err != nil {
		return nil, err
	}

	return &PreheatingStatus{
		TaskID: result.ID,
		Status: provider.PreheatingStatusPending, // default
	}, nil
```

Kraken: error & non-error & PreheatingStatusSuccess

```go
_, err = client.GetHTTPClient(kd.instance.Insecure).Post(url, kd.getCred(), payload, nil)
	if err != nil {
		return nil, err
	}

	return &PreheatingStatus{
		TaskID:     eventID,
		Status:     provider.PreheatingStatusSuccess,
		FinishTime: time.Now().String(),
	}, nil
```